### DISCUSSION TITLE
GLSP Theia client: How to send data to the Theia Output pane from the GLSP Java server?

### DISCUSSION THREAD

**Discussion by mrcalvin** [eclipse-glsp/glsp#1305] [General]
I would like to send results of some analysis performed in the Java GLSP server on a given source-model instance to the Output pane in the Theia client. I checked the project templates and the tutorials, and also the discussions, here. I clearly have missed the obvious?

**Comment by planger** [comment]
For this requirement, you'll need to connect GLSP API with Theia's API. Thus, this needs to go into your Theia integration package (cf. [tasklist-theia](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia/glsp-client/tasklist-theia)) in order to be able get access to Theia's dependency injection container and make this available to your GLSP client package.

A good example of how to do that can be found in the [GLSP-Theia selection module](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/features/select/theia-select-module.ts) and other [GLSP-Theia integration features](https://github.com/eclipse-glsp/glsp-theia-integration/tree/master/packages/theia-integration/src/browser/diagram/features). As you can see in the selection module, it binds the injectable class [`TheiaGLSPSelectionForwarder`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/features/select/theia-glsp-selection-forwarder.ts), which can get stuff from Theia's DI container injected, and make it available via the (`TYPES.ISelectionListener`) key in the GLSP client dependency injection container. Thus, in your GLSP client, you can then get the instance of [`TheiaGLSPSelectionForwarder`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/features/select/theia-glsp-selection-forwarder.ts) with its values from the Theia API injected via `@inject(TYPES.ISelectionListener) ...`. For other platform integrations, an alternative implementation can then be injected.

To hook up the feature module in your diagram, you'll have to add it to your Theia-specific diagram configuration (see e.g. [`TasklistDiagramConfiguration`](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-client/tasklist-theia/src/browser/diagram/tasklist-diagram-configuration.ts)). There you can overwrite `getContainerConfiguration()` and add your feature module to the default list.

There is two patterns for actually implementing such an integration (cf. [`TheiaGLSPSelectionForwarder`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/features/select/theia-glsp-selection-forwarder.ts)). Either you create an interface in the GLSP client (with methods `writeToOutput(...)`) used by other code in your GLSP client code and this interface is then implemented by your Theia-specific class (like with the selection forwarder example). Alternatively, you introduce an action that is then handled by your Theia-specific action handler (like it is done with the [notification module](https://github.com/eclipse-glsp/glsp-theia-integration/tree/master/packages/theia-integration/src/browser/diagram/features/notification)). These actions can then be triggered by any other code of your GLSP client via the action dispatcher and handled by your Theia-specific implementation (see [actions in the docs](https://eclipse.dev/glsp/documentation/actionhandler/)) based on the Theia API.

Now to the Theia-specific implementation:
To write something to Theia's output view, you can get injected the [`OutputChannelManager`](https://github.com/eclipse-theia/theia/blob/master/packages/output/src/browser/output-channel.ts), create a new channel (`getChannel('Your Channel')`) and write to it (`appendLine('Hello World')`). So that part should be rather straightforward.

**Comment by rsoika** [comment]
Hi @mrcalvin , I do not know what you plan in detail but be aware also of the [GLSP Validation feature](https://eclipse.dev/glsp/documentation/validation/) that allows you to validate a model live and in batch mode (triggered from the Tool Palette)
